### PR TITLE
fix: Rework git tag semver resolution (#20083)

### PR DESCRIPTION
**Pull request by PaulSonOfLars**
Testing a potential fix for #20083, where ArgoCD's semver resolution is too lax, which causes non-semver tags to be incorrectly coerced into semver. This then causes unexpected tags to be used when they shouldn't be.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Review by nitishfy — COMMENTED** [review]
can you fix the CI failures?

**Comment by PaulSonOfLars** [comment]
@nitishfy apologies for that, the PR is just aiming to be a draft for now - to see if the silly basic approach would work against the full testsuite (onboarding and spinning up the tests locally is quite painful if you're not a regular contributor!). Clearly it doesn't!

I've now narrowed down on the exact issue, and should be able to make the git semver logic match the helm logic, avoiding any big "breaking" changes.

I'll aim to have that in a proper PR open with tests and the actual fix later tonight, ready for tomorrow's contributors meeting.

File: util/git/client.go [hunk 1]
```diff
@@ -723,6 +731,7 @@ func (m *nativeGitClient) resolveSemverRevision(constraint *semver.Constraints,
 		return ""
 	}
 
+	log.Infof("Semver constraint '%s' resolved to tag '%s', at reference '%s'", revision, maxVersion.Original(), maxVersionHash.String())
```
Comment (rumstead) [inline util/git/client.go:734]: Should we make this a debug log?

File: util/git/client.go [hunk 2]
```diff
@@ -684,18 +679,31 @@ func (m *nativeGitClient) lsRemote(revision string) (string, error) {
...
-// * The revision is "master-branch" only the lsRemote loop will run because that revision is an invalid semver;
-func (m *nativeGitClient) resolveSemverRevision(constraint *semver.Constraints, refs []*plumbing.Reference) string {
+// * The revision is "v0.1.2"/"0.1.2", then this is not a constraint, it's a pinned version - so we fall back to the standard tag matching in the lsRemote loop.
+// * The revision is "v0.1.*"/"0.1.*", and there's a tag matching that constraint, then we find the latest matching version and return its commit hash.
+// * The revision is "v0.1.*"/"0.1.*", and there is *no* tag matching that constraint, then we fall back to the standard tag matching in the lsRemote loop.
+// * The revision is "custom-tag", only the lsRemote loop will run - because that revision is an invalid semver;
+// * The revision is "master-branch", only the lsRemote loop will run because that revision is an invalid semver;
+func (m *nativeGitClient) resolveSemverRevision(revision string, refs []*plumbing.Reference) string {
+	if _, err := semver.NewVersion(revision); err == nil {
```
Comment (rumstead) [inline util/git/client.go:695]: Do we need this check given a a valid semver isn't a constraint and 

```
constraint, err := semver.NewConstraint(revision)
```
should return an error?

**Review by blakepettersson — APPROVED** [review]
I agree with @rumstead's comment, otherwise LGTM. To unblock this error I'd ship this as-is, but I would like to see a follow-up PR which consolidates this with the Helm semver resolution.

**Review by gdsoumya — APPROVED** [review]
LGTM

**Review by gdsoumya — CHANGES_REQUESTED** [review]

**Review by christianh814 — APPROVED** [review]

File: util/git/client_test.go [hunk 3]
```diff
@@ -173,6 +173,132 @@ func Test_ChangedFiles(t *testing.T) {
...
+		expected: mapTagRefs["v1.1.0"],
+	}, {
+		name:     "minor gte constraint",
+		ref:      ">= v1.0.0",
+		expected: mapTagRefs["v1.1.0"],
+	}, {
+		name:     "multiple constraints",
+		ref:      "> v1.0.0 < v1.1.0",
+		expected: mapTagRefs["v1.0.1"],
+	}, {
```
[the remark this answers is not in the document]
Comment (PaulSonOfLars) [reply util/git/client_test.go:253]: Just to clarify - are you asking to update the argo docs, or to add a test that covers this case? Or both? :)

[the remark this answers is not in the document]
Comment (blakepettersson) [reply util/git/client_test.go:253]: I think Soumya would like a test case that exercises this case, with ideally some kind of defined expectation on what this should do.

[the remark this answers is not in the document]
Comment (PaulSonOfLars) [reply util/git/client_test.go]: Gotcha - done. 
Added a tescase for v1.0, which should only resolve to v1.0.
If users want wildcard versions, they should be explicitly defined

[the remark this answers is not in the document]
Comment (PaulSonOfLars) [reply util/git/client_test.go]: Done - but worth noting that these functions are all going to be removed in the follow up PR to unify semver resolution anyway

**Review by nitishfy — APPROVED** [review]
LGTM!

**Review by gdsoumya — APPROVED** [review]

**Comment by blakepettersson** [comment]
@PaulSonOfLars can you create a cherry-pick PR for 2.12 as well?

**Comment by PaulSonOfLars** [comment]
> @PaulSonOfLars can you create a cherry-pick PR for 2.12 as well?

Done - https://github.com/argoproj/argo-cd/pull/20215

**Pull request merged**